File: src/scripts/titanic_model.py
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

print("Step 1: Loading data...")

train_df = pd.read_csv("src/data/train.csv")
test_df = pd.read_csv("src/data/test.csv")
gender_baseline = pd.read_csv("src/data/gender_submission.csv")

print(f"train.csv shape: {train_df.shape}")
print(f"test.csv shape: {test_df.shape}")
print(f"gender_submission.csv shape: {gender_baseline.shape}")

print("\nStep 2: Cleaning and feature engineering...")

# Fill missing values
train_df["Age"].fillna(train_df["Age"].median(), inplace=True)
test_df["Age"].fillna(test_df["Age"].median(), inplace=True)
train_df["Embarked"].fillna(train_df["Embarked"].mode()[0], inplace=True)
test_df["Fare"].fillna(test_df["Fare"].median(), inplace=True)

# Encode categorical features
for df in [train_df, test_df]:
    df["Sex"] = df["Sex"].map({"male": 0, "female": 1})
    df["Embarked"] = df["Embarked"].map({"S": 0, "C": 1, "Q": 2})

# Feature selection
features = ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"]
X = train_df[features]
y = train_df["Survived"]

print("Features used:", features)
print(f"X shape: {X.shape}")
print(f"y shape: {y.shape}")

print("\nStep 3: Train/validation split and model training...")

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Evaluate training and validation accuracy
y_pred_train = model.predict(X_train)
train_acc = accuracy_score(y_train, y_pred_train)
print(f"Training accuracy: {train_acc:.3f}")

y_pred_val = model.predict(X_val)
val_acc = accuracy_score(y_val, y_pred_val)
print(f"Validation accuracy: {val_acc:.3f}")

print("\nStep 4: Predict on test.csv...")

test_X = test_df[features]
test_pred = model.predict(test_X)

# Save predictions
my_test_pred = pd.DataFrame({
    "PassengerId": test_df["PassengerId"],
    "Survived": test_pred
})

print("Sample predictions:")
print(my_test_pred.head())

print("\nStep 5: Compare to gender_submission.csv (assignment baseline)...")

merged = pd.merge(
    my_test_pred,
    gender_baseline,
    on="PassengerId",
    suffixes=("_model", "_gender")
)

match_acc = (merged["Survived_model"] == merged["Survived_gender"]).mean()
print(f"Match rate vs gender_submission.csv: {match_acc:.3f}")

print("\nNote:")
print(" - Training accuracy shows model performance on labeled data.")
print(" - Validation accuracy measures generalization on unseen data.")
print(" - The gender_submission.csv comparison is used as a test reference for this assignment.")
print("\nStep 6: Saving predictions...")

my_test_pred.to_csv("src/data/titanic_predictions.csv", index=False)
print("Saved: src/data/titanic_predictions.csv")

print("\nScript finished successfully.")

File: src/r/titanic_model.R
library(dplyr)
library(ggplot2)
library(readr)
library(tidyr)
library(caret)


cat("Step 1: Loading data...\n")

train <- read.csv("src/data/train.csv")
test  <- read.csv("src/data/test.csv")
gender <- read.csv("src/data/gender_submission.csv")

cat(paste("train.csv shape:", dim(train)[1], "rows,", dim(train)[2], "columns\n"))
cat(paste("test.csv shape:", dim(test)[1], "rows,", dim(test)[2], "columns\n"))

cat("\nStep 2: Cleaning data...\n")

# Fill missing values
train$Age[is.na(train$Age)] <- median(train$Age, na.rm = TRUE)
test$Age[is.na(test$Age)] <- median(test$Age, na.rm = TRUE)
train$Embarked[train$Embarked == ""] <- "S"
test$Fare[is.na(test$Fare)] <- median(test$Fare, na.rm = TRUE)

# Encode categorical variables
train$Sex <- ifelse(train$Sex == "male", 0, 1)
test$Sex  <- ifelse(test$Sex == "male", 0, 1)
train$Embarked <- recode(train$Embarked, "S" = 0, "C" = 1, "Q" = 2)
test$Embarked  <- recode(test$Embarked, "S" = 0, "C" = 1, "Q" = 2)

# Select features
features <- c("Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked")

cat("\nStep 3: Training logistic regression...\n")

# Build logistic regression model
formula <- as.formula(paste("Survived ~", paste(features, collapse = " + ")))
model <- glm(formula, data = train, family = binomial())

# Predict on training set
train$pred_train <- ifelse(predict(model, train, type = "response") > 0.5, 1, 0)
train_acc <- mean(train$pred_train == train$Survived)
cat(paste("Training accuracy:", round(train_acc, 3), "\n"))

cat("\nStep 4: Predicting on test.csv...\n")

test$Survived <- ifelse(predict(model, test, type = "response") > 0.5, 1, 0)

# Compare to gender_submission.csv
merged <- merge(test[, c("PassengerId", "Survived")], gender, by = "PassengerId")
match_acc <- mean(merged$Survived.x == merged$Survived.y)
cat(paste("Match rate vs gender_submission.csv:", round(match_acc, 3), "\n"))

cat("\nStep 5: Saving predictions...\n")
write.csv(test[, c("PassengerId", "Survived")], "src/data/titanic_predictions_r.csv", row.names = FALSE)
cat("Saved: src/data/titanic_predictions_r.csv\n")

cat("\nScript finished successfully.\n")
